File: licensing/sweep_usages_test.go
package licensing_test

import (
	"testing"
	"time"

	lic "github.com/AnoRebel/licensing/licensing"
	"github.com/AnoRebel/licensing/licensing/storage/memory"
)

type sweepClock struct{ now string }

func (c *sweepClock) NowISO() string { return c.now }

// A seat that stops heartbeating must become reclaimable, and one that
// keeps reporting must not — the whole point of last_seen_at.
func TestSweepInactiveUsages(t *testing.T) {
	clk := &sweepClock{now: "2026-06-01T00:00:00.000000Z"}
	st := memory.New(memory.Options{Clock: clk})

	lc, err := st.CreateLicense(lic.LicenseInput{
		LicensableType: "User", LicensableID: "u1",
		LicenseKey: "LIC-AAAA-BBBB-CCCC-DDDD-EEEE-FFFF-GGGG-HHHH",
		Status:     lic.LicenseStatusActive, MaxUsages: 5,
	})
	if err != nil {
		t.Fatal(err)
	}

	mk := func(fp, seen string) *lic.LicenseUsage {
		u, err := st.CreateUsage(lic.LicenseUsageInput{
			LicenseID: lc.ID, Fingerprint: fp,
			Status: lic.UsageStatusActive, RegisteredAt: seen,
		})
		if err != nil {
			t.Fatal(err)
		}
		return u
	}

	fresh := mk("a1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6e7f8a9b0c1d2e3f4a5b6c7d8e9f0a1b2",
		"2026-05-31T23:00:00.000000Z") // 1h ago
	stale := mk("b1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6e7f8a9b0c1d2e3f4a5b6c7d8e9f0a1b2",
		"2026-04-01T00:00:00.000000Z") // ~2 months ago

	// A new usage inherits its liveness clock from registration.
	if fresh.LastSeenAt != "2026-05-31T23:00:00.000000Z" {
		t.Fatalf("last_seen_at should default to registered_at, got %q", fresh.LastSeenAt)
	}

	// Dry run must report without mutating.
	dry, err := lic.SweepInactiveUsages(st, clk, lic.SweepInactiveUsagesOptions{
		InactiveFor: 30 * 24 * time.Hour, DryRun: true,
	})
	if err != nil {
		t.Fatal(err)
	}
	if len(dry.Stale) != 1 || dry.Stale[0] != stale.ID {
		t.Fatalf("dry run should flag exactly the stale seat, got %v", dry.Stale)
	}
	if len(dry.Revoked) != 0 {
		t.Fatalf("dry run must not revoke anything, got %v", dry.Revoked)
	}
	if got, _ := st.GetUsage(stale.ID); got.Status != lic.UsageStatusActive {
		t.Fatal("dry run mutated the seat")
	}

	// Live run revokes only the stale seat.
	res, err := lic.SweepInactiveUsages(st, clk, lic.SweepInactiveUsagesOptions{
		InactiveFor: 30 * 24 * time.Hour,
	})
	if err != nil {
		t.Fatal(err)
	}
	if len(res.Revoked) != 1 || res.Revoked[0] != stale.ID {
		t.Fatalf("expected only the stale seat revoked, got %v", res.Revoked)
	}
	if got, _ := st.GetUsage(stale.ID); got.Status != lic.UsageStatusRevoked {
		t.Fatal("stale seat was not revoked")
	}
	if got, _ := st.GetUsage(fresh.ID); got.Status != lic.UsageStatusActive {
		t.Fatal("fresh seat must survive the sweep")
	}

	// The sweep must leave the same audit trail as an admin revoke.
	page, err := st.ListAudit(lic.AuditLogFilter{Events: []string{"usage.revoked"}}, lic.PageRequest{Limit: 10})
	if err != nil {
		t.Fatal(err)
	}
	if len(page.Items) != 1 {
		t.Fatalf("expected 1 usage.revoked audit row, got %d", len(page.Items))
	}
}

// A non-positive window is programmer misuse and must not silently revoke
// every seat.
func TestSweepInactiveUsages_RejectsNonPositiveWindow(t *testing.T) {
	clk := &sweepClock{now: "2026-06-01T00:00:00.000000Z"}
	st := memory.New(memory.Options{Clock: clk})
	if _, err := lic.SweepInactiveUsages(st, clk, lic.SweepInactiveUsagesOptions{}); err == nil {
		t.Fatal("expected an error for a zero InactiveFor")
	}
}

File: typescript/tests/core/sweep-usages.test.ts
/**
 * Inactivity sweep — seat reclamation.
 *
 * Covers, across every backend:
 *   - A newly registered seat inherits `last_seen_at` from `registered_at`,
 *     so inactivity is measurable without a "never reported" special case.
 *   - A dry run reports stale seats without mutating anything.
 *   - A live sweep revokes only seats past the threshold.
 *   - The sweep emits the same `usage.revoked` audit row as an admin
 *     revoke, so a swept history is not differently shaped.
 *   - A non-positive window is rejected rather than revoking every seat.
 *
 * Mirrors Go's TestSweepInactiveUsages.
 */

import { Database } from 'bun:sqlite';
import { describe, expect, it } from 'bun:test';

import { MemoryStorage } from '@anorebel/licensing/storage/memory';
import { SqliteStorage } from '@anorebel/licensing/storage/sqlite';
import { applyMigrations as applySqliteMigrations } from '@anorebel/licensing/storage/sqlite/migrations';

import { createFixedClock } from '../../src/id.ts';
import { createLicense } from '../../src/license-service.ts';
import type { Storage } from '../../src/storage/types.ts';
import { registerUsage, sweepInactiveUsages } from '../../src/usage-service.ts';

const NOW = '2026-06-01T00:00:00.000000Z';
const DAY_SEC = 24 * 60 * 60;

const FP_FRESH = 'a'.repeat(64);
const FP_STALE = 'b'.repeat(64);

interface Backend {
  readonly name: string;
  readonly make: () => Promise<{ s: Storage; cleanup: () => void }>;
}

const BACKENDS: readonly Backend[] = [
  {
    name: 'memory',
    make: async () => ({
      s: new MemoryStorage({ clock: createFixedClock(NOW) }),
      cleanup: () => undefined,
    }),
  },
  {
    name: 'sqlite',
    make: async () => {
      const db = new Database(':memory:');
      applySqliteMigrations(db);
      return {
        s: new SqliteStorage(db, { clock: createFixedClock(NOW), skipWalPragma: true }),
        cleanup: () => db.close(),
      };
    },
  },
];

for (const backend of BACKENDS) {
  describe(`sweepInactiveUsages (${backend.name})`, () => {
    it('revokes only seats past the inactivity window', async () => {
      const { s, cleanup } = await backend.make();
      try {
        const clock = createFixedClock(NOW);
        const license = await createLicense(s, clock, {
          scope_id: null,
          template_id: null,
          licensable_type: 'User',
          licensable_id: 'u-sweep',
          max_usages: 5,
        });

        // Registration stamps last_seen_at from the registering clock, so
        // each seat is aged by registering it under a different instant.
        const fresh = await registerUsage(s, createFixedClock('2026-05-31T23:00:00.000000Z'), {
          license_id: license.id,
          fingerprint: FP_FRESH,
        });
        const stale = await registerUsage(s, createFixedClock('2026-04-01T00:00:00.000000Z'), {
          license_id: license.id,
          fingerprint: FP_STALE,
        });

        const freshUsage = await s.getUsage(fresh.usage.id);
        expect(freshUsage?.last_seen_at).toBe('2026-05-31T23:00:00.000000Z');

        // Dry run reports without mutating.
        const dry = await sweepInactiveUsages(s, clock, {
          inactiveForSec: 30 * DAY_SEC,
          dryRun: true,
        });
        expect(dry.stale).toEqual([stale.usage.id]);
        expect(dry.revoked).toEqual([]);
        expect((await s.getUsage(stale.usage.id))?.status).toBe('active');

        // Live run revokes only the stale seat.
        const res = await sweepInactiveUsages(s, clock, { inactiveForSec: 30 * DAY_SEC });
        expect(res.revoked).toEqual([stale.usage.id]);
        expect((await s.getUsage(stale.usage.id))?.status).toBe('revoked');
        expect((await s.getUsage(fresh.usage.id))?.status).toBe('active');

        // The sweep must leave the same trail as an admin revoke.
        // NOTE: the TS filter field is `event` (string | string[]); Go's is
        // `Events`. Passing `events` here would be silently ignored as an
        // unknown property and assert nothing.
        const audit = await s.listAudit({ event: 'usage.revoked' }, { limit: 10 });
        expect(audit.items).toHaveLength(1);
      } finally {
        cleanup();
      }
    });

    it('rejects a non-positive window instead of revoking everything', async () => {
      const { s, cleanup } = await backend.make();
      try {
        await expect(
          sweepInactiveUsages(s, createFixedClock(NOW), { inactiveForSec: 0 }),
        ).rejects.toThrow(/positive inactiveForSec/);
      } finally {
        cleanup();
      }
    });
  });
}
